File: Baekjoon/Cpp/Devide_and_Conquer/S1_1992_쿼드트리.cpp
#include <iostream>
#include <string>
using namespace std;

int arr[64][64];

// 시작점 (r,c), 크기 size
void DFS(int r, int c, int size) {
	bool chk = true;
	int num = arr[r][c];
	for (int i = r; i < r + size; ++i) {
		for (int j = c; j < c + size; ++j) {
			if (arr[i][j] != num) {
				chk = false;
				break;
			}
		}
		if (!chk) break;
	}
	if (!chk) {
		cout << "(";
		DFS(r, c, size / 2);
		DFS(r, c + size / 2, size / 2);
		DFS(r + size / 2, c, size / 2);
		DFS(r + size / 2, c + size / 2, size / 2);
		cout << ")";
	}
	else cout << num;
}

int main() {
	ios::sync_with_stdio(false);
	cin.tie(0);
	cout.tie(0);

	// 1992 쿼드트리
	int n;
	string tmp;
	cin >> n;
	for (int i = 0; i < n; ++i){
		cin >> tmp;
		for (int j = 0; j < n; ++j) {
			arr[i][j] = tmp[j]-'0'; // string이라.. char로 자르면 아스키코드가 들어간다..
		}
	}
	
	DFS(0, 0, n);

	return 0;
}

File: Baekjoon/Java/DFS/S1_1992_쿼드트리.java
import java.util.Scanner;

public class Main {
	
	static int N;
	static int[][] arr;
	
	public static void main(String[] args) {
		Scanner sc = new Scanner(System.in);
		N = sc.nextInt();
		arr = new int[N][N];
		for (int i = 0; i < N; ++i) {
			String s = sc.next();
			char[] cs = s.toCharArray();
			for (int j = 0; j < N; ++j) {
				arr[i][j] = cs[j] - '0';
			}
		}
		dfs(0, 0, N);
	}

	private static void dfs(int r, int c, int size) {
		boolean chk = true;
		int num = arr[r][c];
		for (int i = r; i < r + size; ++i) {
			for (int j = c; j < c + size; ++j) {
				if (arr[i][j] != num) {
					chk = false;
					break;
				}
			}
			if(!chk) break;
		}
		if (!chk) {
			System.out.print("(");
			dfs(r, c, size/2);
			dfs(r, c + size/2, size/2);
			dfs(r + size/2, c, size/2);
			dfs(r + size/2, c + size/2, size/2);
			System.out.print(")");
		}
		else System.out.print(num);
		
	}
}
